pkg/advertisement-operator/forge.go

```diff
@@ -0,0 +1,160 @@
+package advertisementOperator
+
+import (
+	advtypes "github.com/liqotech/liqo/apis/sharing/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+)
+
+func forgeVKAffinity() *v1.Affinity {
+	return &v1.Affinity{
+		NodeAffinity: &v1.NodeAffinity{
+			RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+				NodeSelectorTerms: []v1.NodeSelectorTerm{
+					{
+						MatchExpressions: []v1.NodeSelectorRequirement{
+							{
+								Key:      "type",
+								Operator: v1.NodeSelectorOpNotIn,
+								Values:   []string{"virtual-node"},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func forgeVKVolumes(adv *advtypes.Advertisement) []v1.Volume {
+	return []v1.Volume{
+		{
+			Name: "remote-kubeconfig",
+			VolumeSource: v1.VolumeSource{
+				Secret: &v1.SecretVolumeSource{
+					SecretName: adv.Spec.KubeConfigRef.Name,
+				},
+			},
+		},
+		{
+			Name: "virtual-kubelet-crt",
+			VolumeSource: v1.VolumeSource{
+				EmptyDir: &v1.EmptyDirVolumeSource{},
+			},
+		},
+	}
+}
+
+func forgeVKInitContainers(nodeName string, initVKImage string) []v1.Container {
+	return []v1.Container{
+		{
+			Name:  "crt-generator",
+			Image: initVKImage,
+			Command: []string{
+				"/usr/bin/local/kubelet-setup.sh",
+			},
+			Env: []v1.EnvVar{
+				{
+					Name:      "POD_IP",
+					ValueFrom: &v1.EnvVarSource{FieldRef: &v1.ObjectFieldSelector{FieldPath: "status.podIP", APIVersion: "v1"}},
+				},
+				{
+					Name:      "POD_NAME",
+					ValueFrom: &v1.EnvVarSource{FieldRef: &v1.ObjectFieldSelector{FieldPath: "metadata.name", APIVersion: "v1"}},
+				},
+				{
+					Name:  "NODE_NAME",
+					Value: nodeName,
+				},
+			},
+			Args: []string{
+				"/etc/virtual-kubelet/certs",
+			},
+			VolumeMounts: []v1.VolumeMount{
+				{
+					Name:      "virtual-kubelet-crt",
+					MountPath: "/etc/virtual-kubelet/certs",
+				},
+			},
+		},
+	}
+}
+
+func forgeVKContainers(vkImage string, adv *advtypes.Advertisement, nodeName string, vkNamespace string, homeClusterId string) []v1.Container {
+	command := []string{
+		"/usr/bin/virtual-kubelet",
+	}
+
+	args := []string{
+		"--foreign-cluster-id",
+		adv.Spec.ClusterId,
+		"--provider",
+		"kubernetes",
+		"--nodename",
+		nodeName,
+		"--kubelet-namespace",
+		vkNamespace,
+		"--foreign-kubeconfig",
+		"/app/kubeconfig/remote",
+		"--home-cluster-id",
+		homeClusterId,
+	}
+
+	volumeMounts := []v1.VolumeMount{
+		{
+			Name:      "remote-kubeconfig",
+			MountPath: "/app/kubeconfig/remote",
+			SubPath:   "kubeconfig",
+		},
+		{
+			Name:      "virtual-kubelet-crt",
+			MountPath: "/etc/virtual-kubelet/certs",
+		},
+	}
+
+	return []v1.Container{
+		{
+			Name:            "virtual-kubelet",
+			Image:           vkImage,
+			ImagePullPolicy: v1.PullAlways,
```

Comment:
Do we really want the Always policy? Isn't it too impactful on the performance?

Reply:
I just copied what was already there, but I totally agree.